dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -618,18 +665,89 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
     {
         BOOL foregroundHandled = true;
         THROW_IF_FAILED(InvokeAll(payloadLength, payload, &foregroundHandled));
+        THROW_HR_IF(E_UNEXPECTED, !foregroundHandled);
 
-        if (!foregroundHandled)
+        return S_OK;
+    }
+    CATCH_RETURN();
+
+    void PushNotificationManager::Run(winrt::IBackgroundTaskInstance const& taskInstance)
+    {
+        // This function can be triggered by either OS background infrastructure
+        // or by the PushNotificationsLongRunningProcess.
+        bool firstNotificationReceived{ false };
+        {
+            auto lock{ m_lock.lock_shared() };
+            firstNotificationReceived = m_firstNotificationReceived;
+        }
+        if (!firstNotificationReceived)
         {
-            if (!AppModel::Identity::IsPackagedProcess())
             {
-                wil::unique_cotaskmem_string processName;
-                THROW_IF_FAILED(GetCurrentProcessPath(processName));
-                THROW_IF_FAILED(PushNotificationHelpers::ProtocolLaunchHelper(processName.get(), payloadLength, payload));
+                auto lock{ m_lock.lock_exclusive() };
+                m_firstNotificationReceived = true;
+            }
+
+            std::wstring commandLine{ GetCommandLine() };
+            auto pos{ commandLine.find(c_expectedPushServerArgs) };
+            if (pos == std::wstring::npos) // Any launch kind that is not PushNotification
+            {
+                THROW_HR_IF(E_UNEXPECTED, m_foregroundHandlers);
+
+                winrt::guid registeredClsid{ GUID_NULL };
+                THROW_IF_FAILED(PushNotificationHelpers::GetComRegistrationFromRegistry(c_expectedPushServerArgs.data(), registeredClsid));
+
+                auto backgroundTask{ winrt::create_instance<winrt::IBackgroundTask>(registeredClsid, CLSCTX_ALL) };
+                backgroundTask.Run(taskInstance);
+                return;
```

Comment:
bad pattern. Suggest that you structure the code in such a way that we don't return in the middle of a function.

Rationale: If we need to add things later to this, the premature returning could be problematic although it is technically correct now.

Reply:
One way to fix this is to use an if {{} else based on the condition above.

Reply:
See my fix for this.